File: lc_previous/number-of-islands.cpp
class Solution {
private:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        if (r < 0 || r > grid.size() - 1) {
            return;
        }
        
        if (c < 0 || c > grid[0].size()- 1) {
            return;
        }
        
        if (grid[r][c] == '0') {
            return;
        }
        
        grid[r][c] = '0';
        
        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int r = grid.size();
        if (!r) {
            return 0;
        }
        int answer = 0;
        int c = grid[0].size();
        for (int i = 0; i< r; ++i) {
            for (int j=0; j< c; ++j) {
                if (grid[i][j] == '1')
                answer ++;
                dfs(grid, i, j);
            }
        }
        return answer;
    }
};

File: graphs/number_of_islands.py
"""
https://leetcode.com/problems/number-of-islands/
"""

class Solution:
    def traverse(self, grid, i, j):
        if i < 0 or i > len(grid) - 1:
            return

        if j < 0 or j >  len(grid[0]) - 1:
            return

        if grid[i][j] == '0':
            return

        grid[i][j] = '0'

        self.traverse(grid, i- 1, j)
        self.traverse(grid, i + 1, j)
        self.traverse(grid, i, j - 1)
        self.traverse(grid, i, j + 1)

    def numIslands(self, grid: List[List[str]]) -> int:
        answer = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    answer += 1
                    self.traverse(grid, i, j)

        return answer